File: src/TooYoung.Provider.MongoDB/Services/AccountRepository.cs
using System.Threading.Tasks;
using MongoDB.Driver;
using MongoDB.Driver.Linq;
using TooYoung.Core.Models;
using TooYoung.Core.Repository;

namespace TooYoung.Provider.MongoDB.Services
{
    public class AccountRepository : IAccountRepository
    {
        public IMongoCollection<User> Users { get; set; }

        public AccountRepository(IMongoDatabase db)
        {
            Users = db.GetCollection<User>(nameof(User));
        }

        /// <summary>
        /// 根据用户登录名查找用户信息
        /// </summary>
        /// <param name="username"></param>
        /// <returns></returns>
        public async Task<User> FindByUserName(string username)
        {
            return await Users.AsQueryable().FirstOrDefaultAsync(u => u.UserName == username);
        }

        /// <summary>
        /// 插入一个新的用户信息
        /// </summary>
        /// <param name="user"></param>
        /// <returns></returns>
        public async Task<User> Create(User user)
        {
            await Users.InsertOneAsync(user);
            return user;
        }
    }
}

File: src/TooYoung.Provider.Mongo/Repositories/AccountRepository.fs
namespace TooYoung.Provider.Mongo.Repositories
open System
open MongoDB.Driver
open FSharp.Control.Tasks.V2
open TooYoung.Domain.User
open TooYoung.Domain.Repositories
open FsToolkit.ErrorHandling
open FsToolkit.ErrorHandling.Operator.AsyncResult
open TooYoung
open TooYoung.Provider.Mongo
open Utils

type AccountRepository(db: IMongoDatabase) =
    let users = db.GetCollection<User>("User")
    
    let findByUserName (userName:string) =
        users.Find(fun u -> u.UserName = userName).FirstOrDefaultAsync()
        |> Async.AwaitTask
        |> Async.map Option.ofObj

    let findByUserId (userId: Guid) =
        users.Find(fun u -> u.Id = userId).FirstOrDefaultAsync()
        |> Async.AwaitTask
        |> Async.map Option.ofObj

    let create user =
        task {
            do! users.InsertOneAsync(user)
            return user
        }
        |> Async.AwaitTask
        |> Async.map Ok

    let shouldExists findBy p errorMsg =
        findBy p
        |> Async.map (Result.ofSome (fun _ -> Error errorMsg))
    
    let shouldNotExists findBy p errorMsg =
        findBy p
        |> Async.map (Result.ofNone (fun _ -> Error errorMsg))
    
    let getAllUsers () =
        users.AsQueryable().ToListAsync()
        |> Async.AwaitTask
        |> Async.map List.ofSeq
    
    let deleteUser (user: User) =
        users.DeleteOneAsync((fun u -> u.Id = user.Id))
        |> Async.AwaitTask
        |> Async.map (fun _ -> Ok user)

    interface IAccountRepository with
        member this.BeginTransaction() =
            new MongoUnitOfWork(db.Client) :> UnitOfWork
        /// 根据用户登录名查找用户信息
        member this.FindByUserName u = findByUserName u
        member this.FindByIdAsync u = findByUserId u
        member this.GetAllUsers () = getAllUsers()
        member this.DeleteUser user = deleteUser user
        /// 插入一个新的用户信息
        member this.Create (user: User) =
            create user
        /// 更新用户信息
        member this.UpdateAsync (user: User) =
            users.ReplaceOneAsync((fun u -> u.Id = user.Id), user)
            |> Async.AwaitTask
            |> Async.map (fun _ -> Ok user)
